src/state/contract_class_cache.rs

```diff
@@ -0,0 +1,140 @@
+//! # Contract cache system
+//!
+//! The contract caches allow the application to keep some contracts within itself, providing them
+//! efficiently when they are needed.
+//!
+//! The trait `ContractClassCache` provides methods for retrieving and inserting elements into the
+//! cache. It also contains a method to extend the shared cache from an iterator so that it can be
+//! used with the private caches.
+//!
+//! TODO: Right now, it's impossible to implement the issue's `merge_caches` because
+//!   `ContractClassCache::extend` will be called already with the write lock in place. To solve
+//!   this, the lock may be pushed into the cache, but the methods will not be able to have
+//!   `&mut self` as the object.
```

Comment:
Why is it called with the lock in place? Can't it be dropped before the call? What is it that I'm missing?

Reply:
We should update the comment then. I don't think the `TODO` applies anymore.